iscservice/demo/util.c

```diff
@@ -0,0 +1,710 @@
+	struct m0_rpc_item    *item;
+
+	req->cir_sem = &isc_sem;
+
+	item              = &req->cir_fop.f_item;
+	item->ri_session  = req->cir_rpc_sess;
+	item->ri_prio     = M0_RPC_ITEM_PRIO_MID;
+	item->ri_deadline = M0_TIME_IMMEDIATELY;
+	item->ri_nr_sent_max = M0_RPCLIB_MAX_RETRIES;
+	item->ri_ops      = &isc_item_ops;
+
+	m0_fop_get(&req->cir_fop);
+	rc = m0_rpc_post(item);
+	if (rc != 0) {
+		fprintf(stderr, "Failed to send request to %s: rc=%d\n",
+			m0_rpc_conn_addr(req->cir_rpc_sess->s_conn), rc);
+		m0_fop_put(&req->cir_fop);
+	}
+
+	ireqs_list_add_in_order(req);
+
+	return rc;
+}
+
+int m0util_isc_req_send_sync(struct m0util_isc_req *req)
+{
+	struct m0_fop         *reply_fop;
+	struct m0_fop_isc_rep  isc_reply;
+	int                    rc;
+
+	rc = m0_rpc_post_sync(&req->cir_fop, req->cir_rpc_sess, NULL,
+			      M0_TIME_IMMEDIATELY);
+	if (rc != 0) {
+		fprintf(stderr, "Failed to send request to %s: rc=%d\n",
+			m0_rpc_conn_addr(req->cir_rpc_sess->s_conn), rc);
+		return rc;
+	}
+	reply_fop = m0_rpc_item_to_fop(req->cir_fop.f_item.ri_reply);
+	isc_reply = *(struct m0_fop_isc_rep *)m0_fop_data(reply_fop);
+	rc = req->cir_rc = isc_reply.fir_rc;
+	if (rc != 0)
+		fprintf(stderr, "Got error from %s: rc=%d\n",
+			m0_rpc_conn_addr(req->cir_rpc_sess->s_conn), rc);
+	rc = m0_rpc_at_rep_get(&req->cir_isc_fop.fi_ret, &isc_reply.fir_ret,
+			       &req->cir_result);
+	if (rc != 0)
+		fprintf(stderr, "rpc_at_rep_get() from %s failed: rc=%d\n",
+			m0_rpc_conn_addr(req->cir_rpc_sess->s_conn), rc);
+
+	return req->cir_rc == 0 ? rc : req->cir_rc;
+}
+
+static void fop_fini_lock(struct m0_fop *fop)
+{
+	struct m0_rpc_machine *mach = m0_fop_rpc_machine(fop);
+
+	m0_rpc_machine_lock(mach);
+	m0_fop_fini(fop);
+	m0_rpc_machine_unlock(mach);
+}
+
+void m0util_isc_req_fini(struct m0util_isc_req *req)
+{
+	struct m0_fop *reply_fop = NULL;
+
+	if (req->cir_fop.f_item.ri_reply != NULL)
+		reply_fop = m0_rpc_item_to_fop(req->cir_fop.f_item.ri_reply);
+	if (reply_fop != NULL)
+		m0_fop_put_lock(reply_fop);
+	req->cir_fop.f_item.ri_reply = NULL;
+	m0_rpc_at_fini(&req->cir_isc_fop.fi_args);
+	m0_rpc_at_fini(&req->cir_isc_fop.fi_ret);
+	req->cir_fop.f_data.fd_data = NULL;
+	fop_fini_lock(&req->cir_fop);
+}
+
+static int read_conf_params(int idx, const struct param params[], int n)
+{
+	int i;
+	struct conf_params_to_read {
+		const char *name;
+		const char **conf_ptr;
+	} p[] = { { "HA_ENDPOINT_ADDR",      &m0_conf.mc_ha_addr },
+		  { "PROFILE_FID",           &m0_conf.mc_profile },
+		  { "LOCAL_ENDPOINT_ADDR%d", &m0_conf.mc_local_addr },
+		  { "LOCAL_PROC_FID%d",      &m0_conf.mc_process_fid }, };
+	char pname[256];
+
+	for (i = 0; i < ARRAY_SIZE(p); i++) {
+		sprintf(pname, p[i].name, idx);
+		*(p[i].conf_ptr) = param_get(pname, params, n);
+		if (*(p[i].conf_ptr) == NULL) {
+			ERR("%s is not set at %s\n", pname, c0rcfile);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * m0util_init()
+ * init client resources.
+ */
+int m0util_init(int idx)
```

Comment:
better name:
isc_init();

m0util_init(0) is the only use case.

I know we simplify this for the purpose of SAGE cluster.
But this is not needed for a simple Motr deployment and a simple ISC demo.
This makes things complex and not easy to understand.

Reply:
Right, fixed.